### PR TITLE
Streaming data reader

### PR DESCRIPTION
**Pull request by c-j-hughes**
This isn't yet a complete solution to #115 and #99. 

The lowest level support for reading TDS packets as stream and tokenising them lazily is implemented with all tests passing. I've also restructured `AseDataReader` to use an enumerator internally.

Would appreciate thoughts on how to restructure `InternalConnection.InternalExecuteAsync(...)` so that results can evaluated lazily.

### CODE REVIEW

File: src/AdoNetCore.AseClient/Internal/TokenStream.cs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        private void BufferBytes(byte[] buffer, int offset, int count)
+        {
+            // TODO - this should not just loop, burning CPU.
+            if (count > 0)
+            {
+                var remainingBytes = count;
+                var totalReceivedBytes = 0;
+                do
+                {
+                    var receivedBytes = base.Read(buffer, offset + totalReceivedBytes, remainingBytes);
```
Comment (senseibaka) [inline src/AdoNetCore.AseClient/Internal/TokenStream.cs]: One thing I've noticed in other development is that a `socket.Read` can return `0` if the connection breaks (remote closes the connection), rather than throwing a SocketException. *If* that happens here, you get an infinite loop -- might be good to look into it.

File: test/AdoNetCore.AseClient.Tests/Integration/StreamingTests.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            using (var connection = GetConnection())
+            {
+                connection.Execute(CleanUp);
+            }
+        }
+
+        // TODO - a failing test that shows the result sets cannot be streamed.
+        // TODO - use this to implement an IEnumerable socket, an IEnumerable token reader, and make data reader work accordingly.
+        [Test]
+        public void DelayedResultSetTests()
```
Comment (senseibaka) [inline test/AdoNetCore.AseClient.Tests/Integration/StreamingTests.cs]: I tried to run this test locally but it kept on failing because the expected/actual values differed by about 4 seconds. Is that an issue?

File: src/AdoNetCore.AseClient/AseDataReader.cs [hunk 3]
```diff
@@ -9,7 +10,7 @@ namespace AdoNetCore.AseClient
 {
     public sealed class AseDataReader : DbDataReader
     {
-        private readonly TableResult[] _results;
+        private readonly IEnumerator<TableResult> _results;
```
[the remark this answers is not in the document]
Comment (c-j-hughes) [reply src/AdoNetCore.AseClient/AseDataReader.cs]: I had a crack at it, and the issue is that the reader exposes the number of rows in each `TableResult`. It's not clear what you put in there if you don't know. Maybe +1 as each row goes in. I backed it out in the end because I wasn't sure. It would be cool for a very large result set - and do-able if we can suss out what other providers do.

File: src/AdoNetCore.AseClient/AseDataReader.cs [hunk 4]
```diff
@@ -205,7 +201,7 @@ public override long GetChars(int i, long fieldOffset, char[] buffer, int buffer
                 ++cIndex;
             }
 #else
-            Array.Copy(charArray, fieldOffset, buffer, bufferoffset, charsToRead);
+            Array.Copy(charArray, fieldOffset, buffer, bufferOffset, charsToRead);
```
[the remark this answers is not in the document]
Comment (c-j-hughes) [reply src/AdoNetCore.AseClient/AseDataReader.cs]: Yeah I left the ones that weren't `byte[]` alone because you have to do a bit of math or else `Buffer.BlockCopy` corrupts things.

I think you would do `Buffer.BlockCopy(charArray, fieldOffset, buffer, bufferOffset, charsToRead * sizeof(char));` but I didn't want to break random things.

File: src/AdoNetCore.AseClient/Internal/TokenStream.cs [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+            
+            var bytesWrittenToBuffer = 0;
+
+            // If we have some data in the bodyBuffer, we should use that up first.
+            if (_bufferHasBytes)
+            {
+                bytesWrittenToBuffer = GetBufferedBytes(buffer, 0 + offset, count);
+            }
+
+            // If we need more data, let's read if from the network until the buffer is full.
```
[the remark this answers is not in the document]
Comment (c-j-hughes) [reply src/AdoNetCore.AseClient/Internal/TokenStream.cs]: Done.

File: test/AdoNetCore.AseClient.Tests/Integration/StreamingTests.cs (hunk 2 above)
Comment (c-j-hughes) [reply test/AdoNetCore.AseClient.Tests/Integration/StreamingTests.cs]: That's the feature I'm trying to add. Still can't make it work yet though.

File: src/AdoNetCore.AseClient/Internal/TokenStream.cs [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        /// Ensure that the requested bytes have been received. The base <see cref="NetworkStream"/> will
+        /// block until at least one byte is available, but not necessarily wait for all of the requested
+        /// bytes.
+        /// </summary>
+        /// <param name="buffer">The buffer to fill.</param>
+        /// <param name="offset">The position within the buffer to write data into.</param>
+        /// <param name="count">The number of bytes.</param>
+        private void BufferBytes(byte[] buffer, int offset, int count)
+        {
+            // TODO - this should not just loop, burning CPU.
```
[the remark this answers is not in the document]
Comment (c-j-hughes) [reply src/AdoNetCore.AseClient/Internal/TokenStream.cs]: Yeah you're right.

File: src/AdoNetCore.AseClient/AseDataReader.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (senseibaka) [reply src/AdoNetCore.AseClient/AseDataReader.cs]: Good point

File: src/AdoNetCore.AseClient/AseDataReader.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (senseibaka) [reply src/AdoNetCore.AseClient/AseDataReader.cs]: Ah yes, they're chars not bytes

File: src/AdoNetCore.AseClient/AseDataReader.cs [hunk 7]
```diff
@@ -10,24 +10,45 @@ namespace AdoNetCore.AseClient
...
 #if ENABLE_SYSTEM_DATA_COMMON_EXTENSIONS
         private readonly AseCommand _command;
         private DataTable _currentSchemaTable;
 #endif
 
-        internal AseDataReader(IEnumerable<TableResult> results, AseCommand command, CommandBehavior behavior)
+        internal AseDataReader(AseCommand command, CommandBehavior behavior)
         {
-            _results = results.GetEnumerator();
+            _results = new BlockingCollection<TableResult>();
```
Comment (senseibaka) [inline src/AdoNetCore.AseClient/AseDataReader.cs]: I'd say put a comment in saying that the default underlying collection is FIFO (`ConcurrentQueue`)

File: src/AdoNetCore.AseClient/Internal/InternalConnection.cs [hunk 8]
```diff
@@ -507,7 +571,7 @@ private string MakeCommand(string commandText, CommandBehavior behavior)
             return result;
         }
 
-        // TODO - if namedParameters is false, then look for ? characters in the command, and bind the parameters by position.
+        // TODO - if namedParameters is false, then look for ? characters in the DoTheGoodStuff, and bind the parameters by position.
```
Comment (senseibaka) [inline src/AdoNetCore.AseClient/Internal/InternalConnection.cs]: DoTheGoodStuff?

File: src/AdoNetCore.AseClient/Internal/Handler/StreamingDataReaderTokenHandler.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            }
+        }
+
+        private void ReturnCurrent()
+        {
+            if (_current != null)
+            {
+                _dataReader.AddResult(_current);
+
+                if (!_hasFirst)
```
Comment (senseibaka) [inline src/AdoNetCore.AseClient/Internal/Handler/StreamingDataReaderTokenHandler.cs]: I like this idea, it gives an opportunity for an exception to be thrown before the first result.

File: src/AdoNetCore.AseClient/AseDataReader.cs (hunk 7 above)
Comment (c-j-hughes) [reply src/AdoNetCore.AseClient/AseDataReader.cs]: Have backed this out.

File: src/AdoNetCore.AseClient/Internal/InternalConnection.cs (hunk 8 above)
Comment (c-j-hughes) [reply src/AdoNetCore.AseClient/Internal/InternalConnection.cs]: Lame copy error. And fixed.

File: src/AdoNetCore.AseClient/Internal/Handler/StreamingDataReaderTokenHandler.cs (hunk 9 above)
Comment (c-j-hughes) [reply src/AdoNetCore.AseClient/Internal/Handler/StreamingDataReaderTokenHandler.cs]: Have backed this out.

**Comment by c-j-hughes** [comment]
I've backed out the AseDataReader aspects - and some of the changes to InternalConnection. @senseibaka please review.

I think it would be wise to re-run the performance benchmarks, but I'm not convinced my rig is the place to do it. The tests have been taking a long time (hours) and I'm not sure why. Could you run through the benchmark tests before this goes to master?

**Comment by senseibaka** [comment]
Having a look, tests seem to run fine on my computer (takes a couple of minutes, not hours). Only one fails: `ConnectionTests.OpenConnection_ToNonExistantServer_ThrowsAseException`, but only in .net core v1.1

File: src/AdoNetCore.AseClient/AseConnection.cs [hunk 10]
```diff
@@ -539,7 +539,37 @@ public static void ClearPools()
 
         public bool IsCaseSensitive()
         {
-            return _internal?.IsCaseSensitive() ?? false;
+            int result;
```
Comment (senseibaka) [inline src/AdoNetCore.AseClient/AseConnection.cs]: Shouldn't this be a separate PR?

File: src/AdoNetCore.AseClient/AseConnection.cs [hunk 11]
```diff
@@ -539,7 +539,37 @@ public static void ClearPools()
...
+            using (var command = CreateCommand())
+            {
+                command.CommandType = CommandType.Text;
+                command.CommandText = "SELECT value FROM master.dbo.sysconfigures WHERE name = 'default sortorder id'";
+
+                result = (int)command.ExecuteScalar();
+            }
+
+            switch (result)
```
Comment (senseibaka) [inline src/AdoNetCore.AseClient/AseConnection.cs]: `HashSet<int>.Contains(...)` might be more appropriate than a switch

File: src/AdoNetCore.AseClient/AseConnection.cs (hunk 10 above)
Comment (c-j-hughes) [reply src/AdoNetCore.AseClient/AseConnection.cs]: Yeah...

File: src/AdoNetCore.AseClient/AseConnection.cs (hunk 11 above)
Comment (c-j-hughes) [reply src/AdoNetCore.AseClient/AseConnection.cs]: Disagree. If you're motivated by performance:

1. I would hazard that the raw performance of a small switch is faster than looking up within a HashSet. To get access to that constant lookup speed you have to jump down at least one stack frame. The cost of instantiating the HashSet would certainly be more, so that would have to be done statically.
2. The instantiation cost and memory overhead of a static HashSet would be all the time. And this operation is likely to never be called at all.

File: src/AdoNetCore.AseClient/AseConnection.cs (hunk 10 above)
Comment (c-j-hughes) [reply src/AdoNetCore.AseClient/AseConnection.cs]: Will do.

File: src/AdoNetCore.AseClient/AseConnection.cs (hunk 11 above)
Comment (c-j-hughes) [reply src/AdoNetCore.AseClient/AseConnection.cs]: OK. Did the science. 10m iterations of each is pretty consistently 390ms for hashset, and 250ms for switch. Not the best test harness. Also tested binary search of an ordered array which was worst of all at 540ms.

Comment (senseibaka) [reply src/AdoNetCore.AseClient/AseConnection.cs]: 👍
